v7 prep - includes redesigning parts of core, and unifying different types

devoidfury opened the pull request:
The intent of this PR is to do a few things -

* Refactor parts of core to better suit the newer declarative API
* Simplify underlying code where possible
* Unify and remove duplicate implementations of the same underlying type
* Make the API a bit more consistent
* Inline some or all of the XSD schemas for the types we're using as comments, as discussed in #947
* Clean up output values to be conformant to the spec, throw errors instead of silently failing

It's not ready to merge quite yet - I'm still working on things here; so far I've removed more code than I've added, even after adding a lot of comments and a few features. Opening this PR so you can review what's been done so far and start the discussion around these changes.

There may be a few items worth cherry-picking onto v6 - for example, commit 528be93, which fixes issue #507.

dolanmiu commented:
Wow, very well done,

Could you add a list of breaking changes so I can add it to the release notes, like so:

https://github.com/dolanmiu/docx/releases

devoidfury commented:
Yes, I'll make you a list when it's ready. I still need to go over:

- export/
- file/document
- file/drawing
- file/numbering
- file/settings
- file/styles

and a couple other areas first.

devoidfury commented:
I'm wrapping it up, that's good enough for now. I did not get deep into file/drawing -- that should probably be done at some point, but for now we can move ahead without reworking it.

Here's the breaking changes and notes:


New features:

- In several places, as allowed by the spec, Universal Measure units can now be used.
Universal Measure units are strings in the format /-?[0-9]+(\.[0-9]+)?(mm|cm|in|pt|pc|pi)/, for example: "14pt", "20mm", "2in", "2cm".
These include page margins, page size, paragraph indents, and font sizes. These should be experimented with, some of these may not work and need testing -- even if it's technically valid according to the spec.

- Added a bunch of formats to NumberFormat.

- Much more compliant docx files, which should act more consistently in different editors.

- Reduced bundle size slightly.



Breaking changes:

- Renamed enum PageNumberFormat -> NumberFormat.

- Section vertical align enum changed: SectionVerticalAlignValue -> VerticalAlign

- Runtime checks were added for some values. These can now throw errors for invalid arguments in some cases.

- color values must now be in hex (eg "00CCFF") for Underline, Shading, run color, background color, border color

- Paragraph border argument signature changed.
old:
```ts
  new Paragraph({
      border: {
          top: {
              value: "single",
              size: 6,
              color: "red",
          },
      }
  })
```
new:
```ts
  new Paragraph({
      border: {
          top: {
              style: BorderStyle.SINGLE,
              size: 6,
              color: "FF0000",
          },
      }
  })
```

- For TableOfContents, updateFields option in Document must be explicitly set:
```
new Document({
  features: {
      updateFields: true,
  },
  // ...
})
```

- ITableShadingAttributesProperties changed to IShadingAttributesProperties

- TextRun shading option signature changed to be consistent with shading everywhere else.

old:
```ts
  shading: {
      type: ShadingType.REVERSE_DIAGONAL_STRIPE,
      color: "00FFFF",
      fill: "FF0000",
  },
```
new:
```ts
  shading: {
      val: ShadingType.REVERSE_DIAGONAL_STRIPE,
      color: "00FFFF",
      fill: "FF0000",
  },
```

- A great many internal classes were shifted around or removed. Internal methods were removed in several places. These changes should not affect you if you stuck to the public API (what's in the demos and docs)

dolanmiu commented:
Been reviewing it recently

Will let you know!

dolanmiu commented on src/file/core-properties/properties.ts:
```diff
@@ -48,27 +71,39 @@ export class CoreProperties extends XmlComponent {
         );
         if (options.title) {
-            this.root.push(new Title(options.title));
+            this.root.push(new StringContainer("dc:title", options.title));
```
Nice one, using a generic component like this probably makes more sense

dolanmiu commented on src/file/document/body/section-properties/doc-grid/doc-grid.ts:
```diff
@@ -1,13 +0,0 @@
-import { XmlComponent } from "file/xml-components";
-import { DocGridAttributes } from "./doc-grid-attributes";
-
-export class DocumentGrid extends XmlComponent {
```
A lot of deleted files

Waiting to see where these get added in

devoidfury replied:
A lot of things were moved around, a lot of them were duplicates of the same underlying types and were deleted.

dolanmiu commented on src/file/document/body/section-properties/properties/doc-grid.ts:
```diff
@@ -0,0 +1,38 @@
+//     <xsd:attribute name="linePitch" type="ST_DecimalNumber"/>
+//     <xsd:attribute name="charSpace" type="ST_DecimalNumber"/>
+// </xsd:complexType>
+export interface IDocGridAttributesProperties {
```
Ok seems like they got moved around here

dolanmiu commented on src/file/document/body/section-properties/section-properties.ts:
```diff
@@ -43,53 +35,93 @@ export interface ISectionPropertiesOptions {
+// </xsd:group>
+
+export const sectionMarginDefaults = {
+    TOP: "1in",
```
Good to know it doesn't nessesarily need TWIPS

dolanmiu commented on src/file/file.ts:
```diff
@@ -132,18 +134,6 @@ export class File {
         }
-
-        if (options.features) {
-            if (options.features.trackRevisions) {
```
Is this getting added back anywhere?

dolanmiu replied:
Otherwise Demo 60 won't work?

https://github.com/dolanmiu/docx/blob/730e33b1644e541c18aab6413794220eb20edab9/demo/60-track-revisions.ts

devoidfury replied:
src/file/settings/settings.ts

devoidfury replied:
Rather than calling this.settings.addTrackRevisions(), it's passed straight as an option into the Settings object which handles it in the constructor.

dolanmiu commented on src/file/paragraph/properties.ts:
```diff
@@ -148,8 +140,8 @@ export class ParagraphProperties extends IgnoreIfEmptyXmlComponent {
 
-        if (options.bidirectional) {
-            this.push(new Bidirectional());
+        if (options.bidirectional !== undefined) {
```
any reason for this to check for `!== undefined`? is it not more concise to simply `if (options.bidirectional)`

dolanmiu replied:
Sorry, I misunderstood, this can be set to `false` i presume

devoidfury replied:
yes, these can be explicitly set to false.

dolanmiu commented on src/file/styles/style/paragraph-style.ts:
```diff
@@ -26,40 +17,12 @@ export class StyleForParagraph extends Style {
         this.root.push(this.paragraphProperties);
         this.root.push(this.runProperties);
-
-        if (options.basedOn) {
```
Are these not needed? Why are they deleted?

dolanmiu replied:
I can see they were moved to the `Style` class

devoidfury replied:
most of the shared options between the different types of Style classes were moved to the base class

dolanmiu commented on src/file/table/table-cell/table-cell-components.ts:
```diff
@@ -1,78 +1,53 @@
-
-        return this;
+        if (options.top) {
+            this.root.push(new BorderElement("w:top", options.top));
```
Very nice and concise

dolanmiu commented on src/file/table/table-cell/table-cell.ts:
```diff
@@ -1,108 +1,23 @@
             this.root.push(child);
         }
-
-        if (options.width) {
```
Are these properties added back anywhere?

devoidfury replied:
src/file/table/table-cell/table-cell-properties.ts

dolanmiu commented on src/file/xml-components/simple-elements.ts:
```diff
@@ -0,0 +1,65 @@
+import { Attributes, XmlComponent } from "file/xml-components";
+
+import { hpsMeasureValue } from "../values";
```
I like this file, probably cuts like half of the codebase

dolanmiu commented:
@devoidfury Left some comments

Feel free to reply and address, will approve after!

devoidfury commented on docs/usage/text.md:
```diff
@@ -83,7 +83,7 @@ const text = new TextRun({
     text: "shading",
     shading: {
-        type: ShadingType.REVERSE_DIAGONAL_STRIPE,
+        val: ShadingType.REVERSE_DIAGONAL_STRIPE,
```
Sure, will do.

devoidfury commented on src/convenience-functions.ts:
```diff
@@ -12,7 +12,7 @@ export const convertInchesToTwip = (inches: number): number => {
 
 export const uniqueNumericId = (): number => {
-    return currentCount++;
+    return ++currentCount;
```
#507

dolanmiu replied:
Thanks, nice one

devoidfury commented:
They weren't deleted, just moved to src/file/settings/settings.ts

devoidfury commented:
Rather than calling `this.settings.addTrackRevisions()`, it's passed straight as an option into the `Settings` object which handles it in the constructor.

dolanmiu commented:
@devoidfury Does demo 60 still work?

devoidfury commented:
@dolanmiu it appears to - I didn't remove the "trackRevisions" feature

devoidfury commented:
src/file/settings/settings.ts

Rather than calling this.settings.addTrackRevisions(), it's passed straight as an option into the Settings object which handles it in the constructor.

dolanmiu commented:
@devoidfury Thank you for clarifying

dolanmiu commented:
@devoidfury Merged in!
